core/src/apps/cardano/sign_tx/signer.py

```diff
@@ -749,6 +969,125 @@ def _validate_required_signer(
+
+        # Datums and reference scripts are forbidden, see _validate_collateral_return.
+        output_items_count = 2
+        if output.format == CardanoTxOutputSerializationFormat.LEGACY:
+            output_list: HashBuilderList = HashBuilderList(output_items_count)
+            with self.tx_dict.add(TX_BODY_KEY_COLLATERAL_RETURN, output_list):
+                await self._process_legacy_output(
+                    output_list, output, should_show_tokens
+                )
+        elif output.format == CardanoTxOutputSerializationFormat.POST_ALONZO:
+            output_dict: HashBuilderDict[int, Any] = HashBuilderDict(
+                output_items_count, wire.ProcessError("Invalid collateral return")
+            )
+            with self.tx_dict.add(TX_BODY_KEY_COLLATERAL_RETURN, output_dict):
+                await self._process_post_alonzo_output(
+                    output_dict, output, should_show_tokens
+                )
+        else:
+            raise RuntimeError  # should be unreachable
+
+    def _validate_collateral_return(self, output: messages.CardanoTxOutput) -> None:
+        self._validate_output(output)
+
+        address_type = self._get_output_address_type(output)
+        if address_type not in addresses.ADDRESS_TYPES_PAYMENT_KEY:
+            raise wire.ProcessError("Invalid collateral return")
+
+        if (
+            output.datum_hash is not None
+            or output.inline_datum_size > 0
+            or output.reference_script_size > 0
+        ):
+            raise wire.ProcessError("Invalid collateral return")
+
+    async def _show_collateral_return_init(
+        self, output: messages.CardanoTxOutput
+    ) -> None:
+        # We don't display missing datum warning since datums are forbidden.
+
+        if output.asset_groups_count > 0:
+            await layout.warn_tx_output_contains_tokens(
+                self.ctx, is_collateral_return=True
+            )
+
+        if output.address_parameters is not None:
+            address = addresses.derive_human_readable(
+                self.keychain,
+                output.address_parameters,
+                self.msg.protocol_magic,
+                self.msg.network_id,
+            )
+            await self._show_output_credentials(
+                output.address_parameters,
+            )
+        else:
+            assert output.address is not None  # _validate_output
+            address = output.address
+
+        await layout.confirm_sending(
+            self.ctx,
+            output.amount,
+            address,
+            "Collateral return",
+            self.msg.network_id,
+        )
+
+    def _should_show_collateral_return_init(
+        self, output: messages.CardanoTxOutput
+    ) -> bool:
+        if self.msg.total_collateral is None:
+            return True
+
+        if output.address_parameters is not None:  # change output
+            if not should_show_credentials(output.address_parameters):
+                return False
+
```

Comment:
We use `_is_change_output` to `determine what message to show to the user when confirming sending` (as stated in its docstring) and we override it in the `PlutusSigner` (since there are no change outputs, only device-owned outputs), so using it in `_is_simple_change_output` could affect the displaying logic.

I don't see harm in keeping the current implementation, because the condition `output.address_parameters is not None` is really simple and `_is_change_output` is used for something else.